common/tier4_autoware_utils/include/tier4_autoware_utils/trajectory/trajectory.hpp

```diff
@@ -341,6 +364,120 @@ double calcArcLength(const T & points)
+{
+  validateNonEmpty(points);
+
+  if (points.size() - 1 < src_idx) {
+    throw std::out_of_range("Invalid source index");
+  }
+
+  if (points.size() == 1) {
+    return {};
+  }
+
+  if (offset < 0.0) {
+    auto reverse_points = points;
+    std::reverse(reverse_points.begin(), reverse_points.end());
+    return calcLongitudinalOffsetPoint(
+      reverse_points, reverse_points.size() - src_idx - 1, -offset);
+  }
+
+  double dist_sum = 0.0;
+
+  for (size_t i = src_idx; i < points.size() - 1; ++i) {
+    const auto & p_front = getPoint(points.at(i));
+    const auto & p_back = getPoint(points.at(i + 1));
+
+    const auto dist_segment = calcDistance2d(p_front, p_back);
+    dist_sum += dist_segment;
+
+    const auto dist_res = offset - dist_sum;
+    if (dist_res <= 0.0) {
+      return calcInterpolatePoint(p_back, p_front, std::abs(dist_res / dist_segment));
+    }
+  }
+
+  return {};
+}
+
+/**
+ * @brief calculate the point offset from source point along the trajectory (or path)
+ * @param points points of trajectory, path, ...
+ * @param src_point source point
+ * @param offset length of offset from source point
+ * @return offset point
+ */
+template <class T>
+boost::optional<geometry_msgs::msg::Point> calcLongitudinalOffsetPoint(
+  const T & points, const geometry_msgs::msg::Point & src_point, const double offset)
+{
+  validateNonEmpty(points);
+
+  if (offset < 0.0) {
+    auto reverse_points = points;
+    std::reverse(reverse_points.begin(), reverse_points.end());
+    return calcLongitudinalOffsetPoint(reverse_points, src_point, -offset);
+  }
+
+  const size_t src_seg_idx = findNearestSegmentIndex(points, src_point);
+  const double signed_length_src_offset =
+    calcLongitudinalOffsetToSegment(points, src_seg_idx, src_point);
+
+  return calcLongitudinalOffsetPoint(points, src_seg_idx, offset + signed_length_src_offset);
+}
+
+/**
+ * @brief calculate the point offset from source point along the trajectory (or path)
+ * @param seg_idx segment index of point at beginning of length
+ * @param p_target point to be inserted
+ * @param points output points of trajectory, path, ...
+ * @return index of insert point
+ */
+template <class T>
+inline size_t insertTargetPoint(
+  const size_t seg_idx, const geometry_msgs::msg::Point & p_target, T & points,
+  const double overlap_threshold = 1e-3)
+{
+  validateNonEmpty(points);
+
+  const auto p_front = getPoint(points.at(seg_idx));
+  const auto p_back = getPoint(points.at(seg_idx + 1));
+  const auto yaw = calcAzimuthAngle(p_front, p_back);
+
+  validateNonSharpAngle(p_front, p_target, p_back);
+
+  const auto overlap_with_front = calcDistance2d(p_target, p_front) < overlap_threshold;
+  const auto overlap_with_back = calcDistance2d(p_target, p_back) < overlap_threshold;
+
+  geometry_msgs::msg::Pose target_pose;
+  target_pose.position = p_target;
+  target_pose.orientation = createQuaternionFromYaw(yaw);
+
+  auto p_insert = points.at(seg_idx);
```

Comment:
velocity, acceleration, ... should be interpolated? What do you think.

Reply:
Perhaps it's better to interpolate velocity, acceleration, and others outside of this function

Reply:
> Perhaps it's better to interpolate velocity, acceleration, and others outside of this function

I think so too.